http/websocket/handler.go

```diff
@@ -0,0 +1,133 @@
+			if err != nil {
+				if websocket.IsCloseError(err, websocket.CloseGoingAway,
+					websocket.CloseNormalClosure,
+					websocket.CloseNoStatusReceived,
+					websocket.CloseAbnormalClosure) {
+					logger.Error(fmt.Sprintf("[websocket.Handler] %s closed abruptly: %v", conn.Identifier, err))
+					metrics.Increment("batches_read_total", fmt.Sprintf("status=failed,reason=closeerror,conn_group=%s", conn.Identifier.Group))
+					break
+				}
+				metrics.Increment("batches_read_total", fmt.Sprintf("status=failed,reason=unknown,conn_group=%s", conn.Identifier.Group))
+				logger.Error(fmt.Sprintf("[websocket.Handler] reading message failed. Unknown failure for %s: %v", conn.Identifier, err)) //no connection issue here
+				break
+			}
+
+			timeConsumed := time.Now()
+			metrics.Count("events_rx_bytes_total", len(message), fmt.Sprintf("conn_group=%s", conn.Identifier.Group))
+			payload := &pb.EventRequest{}
+
+			d, s := h.getDeserializerSerializer(messageType)
+			if err := d.Deserialize(message, payload); err != nil {
+				logger.Error(fmt.Sprintf("[websocket.Handler] reading message failed for %s: %v", conn.Identifier, err))
+				metrics.Increment("batches_read_total", fmt.Sprintf("status=failed,reason=serde,conn_group=%s", conn.Identifier.Group))
+				writeBadRequestResponse(conn, s, messageType, err)
+				continue
+			}
+			metrics.Increment("batches_read_total", fmt.Sprintf("status=success,conn_group=%s", conn.Identifier.Group))
+			metrics.Count("events_rx_total", len(payload.Events), fmt.Sprintf("conn_group=%s", conn.Identifier.Group))
+			collector.Collect(r.Context(), &collection.CollectRequest{
+				ConnectionIdentifier: &conn.Identifier,
+				TimeConsumed:         timeConsumed,
+				EventRequest:         payload,
+			})
+			writeSuccessResponse(conn, s, messageType, payload.ReqGuid)
+		}
+	}
+
+}
+
+func writeSuccessResponse(conn connection.Conn, serializer serialization.Serializer, messageType int, requestGUID string) {
+	response := &pb.EventResponse{
+		Status:   pb.Status_SUCCESS,
+		Code:     pb.Code_OK,
+		SentTime: time.Now().Unix(),
+		Reason:   "",
+		Data: map[string]string{
+			"req_guid": requestGUID,
+		},
+	}
+	success, _ := serializer.Serialize(response)
+	conn.WriteMessage(messageType, success)
+}
+
+func writeBadRequestResponse(conn connection.Conn, serializer serialization.Serializer, messageType int, err error) {
+	response := &pb.EventResponse{
+		Status:   pb.Status_ERROR,
+		Code:     pb.Code_BAD_REQUEST,
+		SentTime: time.Now().Unix(),
+		Reason:   fmt.Sprintf("cannot deserialize request: %s", err),
+		Data:     nil,
+	}
+
+	failure, _ := serializer.Serialize(response)
+	conn.WriteMessage(messageType, failure)
+}
+
+func (h *Handler) getDeserializerSerializer(messageType int) (deserialization.Deserializer, serialization.Serializer) {
+	switch messageType {
+	case websocket.BinaryMessage:
+		return deserialization.ProtoDeserilizer(), serialization.ProtoDeserilizer()
+	case websocket.TextMessage:
```

Comment:
Even I was thinking in the direction of extensibility but then I thought that even if we add something in the future these can act as default data formats for each message type, like proto for binary message type and JSON for text. With respect to backward compatibility, we will assume these data formats when nothing is specified. 
Also, I think we can restrict a particular connection with a particular data format in case we are using subprotocol handshake. WDYT?